File: mindspore/lite/src/train/lr_scheduler.cc
/**
 * Copyright 2020 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "include/train/lr_scheduler.h"
#include <sys/stat.h>
#include <algorithm>
#include <utility>
#include <vector>
#include <iostream>
#include <fstream>
#include <memory>
#include "include/errorcode.h"
#include "include/train/train_session.h"
#include "src/common/utils.h"
#include "src/tensor.h"

namespace mindspore {
namespace lite {

int MultiplicativeLRLambda(float *lr, int epoch, void *lr_cb_data) {
  if ((lr == nullptr) || (lr_cb_data == nullptr)) {
    MS_LOG(ERROR) << "nullptr passed as input to MultiplicativeLRLambda";
    return DONT_UPDATE_LR;
  }
  float mult = *(static_cast<float *>(lr_cb_data));
  *lr = *lr * mult;
  return UPDATE_LR;
}

int StepLRLambda(float *lr, int epoch, void *lr_cb_data) {
  if ((lr == nullptr) || (lr_cb_data == nullptr)) {
    MS_LOG(ERROR) << "nullptr passed as input to MultiplicativeLRLambda";
    return DONT_UPDATE_LR;
  }
  struct StepLRLambda *step_lr_data = (static_cast<struct StepLRLambda *>(lr_cb_data));
  if (((epoch + 1) % step_lr_data->step_size) == 0) {
    *lr = *lr * step_lr_data->gamma;
    return UPDATE_LR;
  }
  return DONT_UPDATE_LR;
}

LRScheduler::LRScheduler(LR_Lambda lambda_func, void *lr_cb_data, int step)
    : lambda_func_(lambda_func), lr_data_(lr_cb_data), step_(step) {}

int LRScheduler::EpochEnd(const session::TrainLoopCallBackData &cb_data) {
  if (((cb_data.epoch_ + 1) % step_) == 0) {
    float lr = cb_data.session_->GetLearningRate();
    int update = lambda_func_(&lr, cb_data.epoch_, lr_data_);
    if (update == UPDATE_LR) {
      int ret = cb_data.session_->SetLearningRate(lr);
      if (ret != RET_OK) {
        MS_LOG(ERROR) << "Error setting Leraning rate in train session";
        return mindspore::session::RET_EXIT;
      }
    }
  }
  return mindspore::session::RET_CONTINUE;
}

}  // namespace lite
}  // namespace mindspore

File: model_zoo/research/nlp/skipgram/src/lr_scheduler.py
# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Learning rate utilities."""

import numpy as np


def exp_decay_lr(learning_rate, decay_rate, decay_step, total_step, is_stair=False):
    """lr[i] = learning_rate∗pow(decay_rate, i / decay_step)
    """
    if is_stair:
        lrs = learning_rate * np.power(decay_rate, np.floor(np.arange(total_step) / decay_step))
    else:
        lrs = learning_rate * np.power(decay_rate, np.arange(total_step) / decay_step)
    return lrs.astype(np.float32)


def poly_decay_lr(learning_rate, end_learning_rate, decay_step, total_step, power, update_decay_step=False):
    """polynomial decay learning rate
    """
    lrs = []
    if update_decay_step:
        for step in range(total_step):
            tmp_decay_step = max(decay_step, decay_step * np.ceil(step / decay_step))
            lrs.append((learning_rate - end_learning_rate) * np.power(1 - step / tmp_decay_step, power)
                       + end_learning_rate)
    else:
        for step in range(total_step):
            step = min(step, decay_step)
            lrs.append((learning_rate - end_learning_rate) * np.power(1 - step / decay_step, power) + end_learning_rate)
    lrs = np.array(lrs)
    return lrs.astype(np.float32)
